### ISSUE TITLE
Sequence contains no elements

### ISSUE THREAD

**Issue by untiedshoes** [nopSolutions/nopCommerce#5746]
nopCommerce version: 4.40.x

Steps to reproduce the problem: We're getting a lot of 'Sequence contains no elements' errors in the logs, relating to '/catalog/getcatalogsubcategories' - we're using the ajax main menu.

`System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Nop.Web.Factories.CatalogModelFactory.PrepareSubCategoriesAsync(Int32 id)
   at Nop.Web.Controllers.CatalogController.GetCatalogSubCategories(Int32 id)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at WebMarkupMin.AspNetCore5.WebMarkupMinMiddlewareBase.Invoke(HttpContext context)
   at StackExchange.Profiling.MiniProfilerMiddleware.Invoke(HttpContext context) in C:\projects\dotnet\src\MiniProfiler.AspNetCore\MiniProfilerMiddleware.cs:line 125
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Nop.Services.Authentication.AuthenticationMiddleware.InvokeAsync(HttpContext context)
   at FoxNetSoft.Plugin.Misc.CDNConnector.Infrastructure.CDNConnectorMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Nop.Services.Installation.InstallUrlMiddleware.InvokeAsync(HttpContext context, IWebHelper webHelper)
   at Nop.Services.Common.KeepAliveMiddleware.InvokeAsync(HttpContext context, IWebHelper webHelper)
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)`

![Screen Shot 2021-06-25 at 13 32 29](https://user-images.githubusercontent.com/12123259/123425273-d8544700-d5b9-11eb-8411-b58fe9038743.png)

**Comment by skoshelev** [comment]
I'm sorry we haven't found a way to replicate the problem in our testing environment. There is a theoretical possibility of such a problem, but only when changing the formation of menu items in the **\wwwroot\js\public.menu.js** file (see **topMenuCategoryLine** and **mobileMenuCategoryLine** functions)

Also, to suppress this error, you may change the code of **CatalogModelFactory.PrepareSubCategoriesAsync** method to this one

```
var doc = await PrepareCategoryXmlDocumentAsync();

var model = from xe in doc.Descendants("CategorySimpleModel")
			where xe.XPathSelectElement("Id").Value == id.ToString()
			select xe;

var element = model.FirstOrDefault();

if (element == null)
	return new List<CategorySimpleModel>();

var models = from xe in element.XPathSelectElements("SubCategories/CategorySimpleModel")
			 select GetCategorySimpleModel(xe);

return models.ToList();
```

Closed #5746

**Comment by untiedshoes** [comment]
@skoshelev

Thanks, I'll implement that, and see how we go :)
